File: backend/app/services/payment_service.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy.orm import selectinload

from app.constants import DELIVERY_FEE_JMD, PAYMENT_ELIGIBLE_STATUS, PAYMENT_METHODS
from app.extensions import db
from app.models.package import Package
from app.models.payment import PaymentCheckout, PaymentCheckoutItem
from app.models.user import User
from app.services.delivery_request_service import resolve_delivery_request_for_payment
from app.services.package_service import add_package_event
from app.utils.datetime_format import utc_isoformat


def _decimal(value) -> Decimal:
    return Decimal(str(value)).quantize(Decimal("0.01"))


def generate_invoice_number() -> str:
    today = datetime.utcnow().strftime("%Y%m%d")
    prefix = f"PB-INV-{today}-"
    count = (
        PaymentCheckout.query.filter(PaymentCheckout.invoice_number.like(f"{prefix}%")).count() + 1
    )
    return f"{prefix}{count:04d}"


def get_package_checkout_item(package: Package) -> PaymentCheckoutItem | None:
    return PaymentCheckoutItem.query.filter_by(package_id=package.id).first()


def package_payment_summaries_for_packages(packages: list[Package]) -> dict[str, dict]:
    from app.constants import PAYMENT_METHOD_LABELS

    if not packages:
        return {}
    package_ids = [pkg.id for pkg in packages]
    items = (
        PaymentCheckoutItem.query.filter(PaymentCheckoutItem.package_id.in_(package_ids))
        .options(
            selectinload(PaymentCheckoutItem.checkout).selectinload(PaymentCheckout.recorded_by)
        )
        .all()
    )
    summaries: dict[str, dict] = {}
    for item in items:
        checkout = item.checkout
        if not checkout:
            continue
        summaries[str(item.package_id)] = {
            "checkout_id": str(checkout.id),
            "invoice_number": checkout.invoice_number,
            "amount_jmd": float(item.amount_jmd),
            "method": checkout.method,
            "method_label": PAYMENT_METHOD_LABELS.get(checkout.method, checkout.method),
            "reference": checkout.reference,
            "notes": checkout.notes,
            "recorded_by_name": checkout.recorded_by.full_name if checkout.recorded_by else None,
            "recorded_at": utc_isoformat(checkout.recorded_at),
        }
    return summaries


def package_payment_summary(package: Package) -> dict | None:
    from app.constants import PAYMENT_METHOD_LABELS

    item = get_package_checkout_item(package)
    if not item or not item.checkout:
        return None
    checkout = item.checkout
    return {
        "checkout_id": str(checkout.id),
        "invoice_number": checkout.invoice_number,
        "amount_jmd": float(item.amount_jmd),
        "method": checkout.method,
        "method_label": PAYMENT_METHOD_LABELS.get(checkout.method, checkout.method),
        "reference": checkout.reference,
        "notes": checkout.notes,
        "recorded_by_name": checkout.recorded_by.full_name if checkout.recorded_by else None,
        "recorded_at": utc_isoformat(checkout.recorded_at),
    }


def list_customer_checkouts(customer: User, limit: int = 100) -> list[PaymentCheckout]:
    return (
        PaymentCheckout.query.filter_by(customer_id=customer.id)
        .order_by(PaymentCheckout.recorded_at.desc())
        .limit(limit)
        .all()
    )


def list_customer_packages(customer: User, limit: int = 100) -> list[Package]:
    return (
        Package.query.options(
            selectinload(Package.customer),
            selectinload(Package.shipment),
        )
        .filter_by(customer_id=customer.id)
        .filter(Package.status != "unidentified")
        .order_by(Package.received_at.desc(), Package.tracking_number.desc())
        .limit(limit)
        .all()
    )


def compute_customer_billing_summary(packages: list[Package]) -> dict:
    total_due_jmd = Decimal("0")
    ready_count = 0
    paid_count = 0

    for pkg in packages:
        if (
            pkg.status == PAYMENT_ELIGIBLE_STATUS
            and pkg.billing_status == "ready"
            and pkg.total_due_jmd is not None
        ):
            total_due_jmd += pkg.total_due_jmd
            ready_count += 1
        elif pkg.billing_status == "paid":
            paid_count += 1

    return {
        "total_due_jmd": float(total_due_jmd.quantize(Decimal("0.01"))),
        "ready_count": ready_count,
        "paid_count": paid_count,
        "package_count": len(packages),
        "currency": "JMD",
    }


def _validate_checkout_packages(
    customer: User,
    package_ids: list,
) -> list[Package]:
    if not package_ids:
        raise ValueError("Select at least one package to checkout")

    if len(package_ids) > 50:
        raise ValueError("Cannot checkout more than 50 packages at once")

    packages: list[Package] = []
    seen: set[str] = set()

    for raw_id in package_ids:
        pid = str(raw_id)
        if pid in seen:
            continue
        seen.add(pid)

        package = Package.query.filter_by(id=raw_id, customer_id=customer.id).first()
        if not package:
            raise ValueError("One or more packages were not found for this customer")
        if package.status != PAYMENT_ELIGIBLE_STATUS:
            raise ValueError(
                f"{package.tracking_number} must be ready for pickup before payment"
            )
        if package.billing_status != "ready":
            raise ValueError(
                f"{package.tracking_number} is not ready for payment ({package.billing_status})"
            )
        if package.total_due_jmd is None:
            raise ValueError(f"{package.tracking_number} has no bill amount")
        if get_package_checkout_item(package):
            raise ValueError(f"{package.tracking_number} is already paid")
        packages.append(package)

    if not packages:
        raise ValueError("Select at least one package to checkout")

    return packages


def record_payment_checkout(
    customer: User,
    package_ids: list,
    *,
    method: str,
    recorded_by: User,
    reference: str | None = None,
    notes: str | None = None,
    processing_fee_jmd: float | None = None,
    include_delivery_fee: bool = False,
) -> PaymentCheckout:
    if method not in PAYMENT_METHODS:
        raise ValueError("Invalid payment method")

    packages = _validate_checkout_packages(customer, package_ids)
    delivery_request, delivery_fee = resolve_delivery_request_for_payment(customer, packages)
    if delivery_fee <= 0 and include_delivery_fee:
        delivery_fee = DELIVERY_FEE_JMD
    total = Decimal("0")
    line_amounts: list[tuple[Package, Decimal]] = []

    for package in packages:
        amount = _decimal(package.total_due_jmd)
        if amount <= 0:
            raise ValueError(f"{package.tracking_number} has an invalid bill amount")
        total += amount
        line_amounts.append((package, amount))

    if delivery_fee > 0:
        total += delivery_fee

    processing_fee = Decimal("0")
    if processing_fee_jmd is not None:
        processing_fee = _decimal(processing_fee_jmd)
        if processing_fee < 0:
            raise ValueError("Processing fee cannot be negative")
        if processing_fee > 0:
            total += processing_fee

    checkout = PaymentCheckout(
        customer_id=customer.id,
        invoice_number=generate_invoice_number(),
        total_jmd=total,
        method=method,
        reference=(reference or "").strip() or None,
        notes=(notes or "").strip() or None,
        recorded_by_id=recorded_by.id,
        recorded_at=datetime.utcnow(),
        delivery_request_id=delivery_request.id if delivery_request else None,
        delivery_fee_jmd=delivery_fee if delivery_fee > 0 else None,
        processing_fee_jmd=processing_fee if processing_fee > 0 else None,
    )
    db.session.add(checkout)
    db.session.flush()

    method_label = method.replace("_", " ")
    for package, amount in line_amounts:
        db.session.add(
            PaymentCheckoutItem(
                checkout_id=checkout.id,
                package_id=package.id,
                amount_jmd=amount,
            )
        )
        package.billing_status = "paid"
        package.updated_at = datetime.utcnow()
        add_package_event(
            package,
            package.status,
            f"Payment recorded ({method_label}) — invoice {checkout.invoice_number}",
        )

    db.session.commit()
    return checkout


def record_package_payment(
    package: Package,
    *,
    method: str,
    recorded_by: User,
    reference: str | None = None,
    notes: str | None = None,
) -> PaymentCheckout:
    """Single-package checkout convenience wrapper."""
    return record_payment_checkout(
        package.customer,
        [str(package.id)],
        method=method,
        recorded_by=recorded_by,
        reference=reference,
        notes=notes,
    )

File: server/src/services/paymentService.ts
import { and, desc, eq, inArray, like, ne, sql } from 'drizzle-orm'
import { PAYMENT_ELIGIBLE_STATUS, PAYMENT_METHODS } from '../constants.js'
import {
  packages,
  paymentCheckoutItems,
  paymentCheckouts,
  users,
  type UserRow,
} from '../db/schema/index.js'
import { db } from '../db/index.js'
import { packagePaymentSummary as packagePaymentSummaryDict } from '../lib/serializers/payment.js'
import { resolveDeliveryRequestForPayment } from './deliveryRequestService.js'
import { addPackageEvent } from './packageService.js'

type PackageRow = typeof packages.$inferSelect
type CheckoutItemRow = typeof paymentCheckoutItems.$inferSelect
type CheckoutRow = typeof paymentCheckouts.$inferSelect

export type PackageCheckoutItemWithCheckout = CheckoutItemRow & {
  checkout: CheckoutRow | null
}

function decimal(value: unknown): number {
  return Math.round(parseFloat(String(value)) * 100) / 100
}

export async function generateInvoiceNumber(): Promise<string> {
  const today = new Date().toISOString().slice(0, 10).replace(/-/g, '')
  const prefix = `PB-INV-${today}-`
  const rows = await db
    .select({ id: paymentCheckouts.id })
    .from(paymentCheckouts)
    .where(like(paymentCheckouts.invoiceNumber, `${prefix}%`))
  const count = rows.length + 1
  return `${prefix}${String(count).padStart(4, '0')}`
}

export async function getPackageCheckoutItem(
  pkg: PackageRow,
): Promise<PackageCheckoutItemWithCheckout | null> {
  const [item] = await db
    .select()
    .from(paymentCheckoutItems)
    .where(eq(paymentCheckoutItems.packageId, pkg.id))
    .limit(1)
  if (!item) return null

  const [checkout] = await db
    .select()
    .from(paymentCheckouts)
    .where(eq(paymentCheckouts.id, item.checkoutId))
    .limit(1)

  return { ...item, checkout: checkout ?? null }
}

export async function packagePaymentSummariesForPackages(
  pkgs: PackageRow[],
): Promise<Record<string, Record<string, unknown>>> {
  if (!pkgs.length) return {}

  const packageIds = pkgs.map((p) => p.id)
  const items = await db
    .select()
    .from(paymentCheckoutItems)
    .where(inArray(paymentCheckoutItems.packageId, packageIds))

  const summaries: Record<string, Record<string, unknown>> = {}
  for (const item of items) {
    const [checkout] = await db
      .select()
      .from(paymentCheckouts)
      .where(eq(paymentCheckouts.id, item.checkoutId))
      .limit(1)
    if (!checkout) continue

    let recordedBy: UserRow | null = null
    if (checkout.recordedById) {
      ;[recordedBy] = await db.select().from(users).where(eq(users.id, checkout.recordedById)).limit(1)
    }

    summaries[item.packageId] = packagePaymentSummaryDict(item, checkout, recordedBy ?? null)
  }
  return summaries
}

export async function packagePaymentSummary(
  pkg: PackageRow,
): Promise<Record<string, unknown> | null> {
  const item = await getPackageCheckoutItem(pkg)
  if (!item?.checkout) return null

  let recordedBy: UserRow | null = null
  if (item.checkout.recordedById) {
    ;[recordedBy] = await db.select().from(users).where(eq(users.id, item.checkout.recordedById)).limit(1)
  }

  return packagePaymentSummaryDict(item, item.checkout, recordedBy ?? null)
}

export async function listCustomerCheckouts(customer: UserRow, limit = 100): Promise<CheckoutRow[]> {
  return db
    .select()
    .from(paymentCheckouts)
    .where(eq(paymentCheckouts.customerId, customer.id))
    .orderBy(desc(paymentCheckouts.recordedAt))
    .limit(limit)
}

export async function listCustomerPackages(customer: UserRow, limit = 100): Promise<PackageRow[]> {
  return db
    .select()
    .from(packages)
    .where(and(eq(packages.customerId, customer.id), ne(packages.status, 'unidentified')))
    .orderBy(desc(packages.receivedAt), desc(packages.trackingNumber))
    .limit(limit)
}

export async function computeCustomerBillingSummary(pkgs: PackageRow[]) {
  let totalDueJmd = 0
  let readyCount = 0
  let paidCount = 0

  for (const pkg of pkgs) {
    if (
      pkg.status === PAYMENT_ELIGIBLE_STATUS &&
      pkg.billingStatus === 'ready' &&
      pkg.totalDueJmd != null
    ) {
      totalDueJmd += parseFloat(pkg.totalDueJmd)
      readyCount += 1
    } else if (pkg.billingStatus === 'paid') {
      paidCount += 1
    }
  }

  return {
    total_due_jmd: Math.round(totalDueJmd * 100) / 100,
    ready_count: readyCount,
    paid_count: paidCount,
    package_count: pkgs.length,
    currency: 'JMD',
  }
}

async function validateCheckoutPackages(
  customer: UserRow,
  packageIds: unknown[],
): Promise<PackageRow[]> {
  if (!packageIds.length) throw new Error('Select at least one package to checkout')
  if (packageIds.length > 50) throw new Error('Cannot checkout more than 50 packages at once')

  const pkgs: PackageRow[] = []
  const seen = new Set<string>()

  for (const rawId of packageIds) {
    const pid = String(rawId)
    if (seen.has(pid)) continue
    seen.add(pid)

    const [pkg] = await db
      .select()
      .from(packages)
      .where(and(eq(packages.id, pid), eq(packages.customerId, customer.id)))
      .limit(1)
    if (!pkg) throw new Error('One or more packages were not found for this customer')
    if (pkg.status !== PAYMENT_ELIGIBLE_STATUS) {
      throw new Error(`${pkg.trackingNumber} must be ready for pickup before payment`)
    }
    if (pkg.billingStatus !== 'ready') {
      throw new Error(`${pkg.trackingNumber} is not ready for payment (${pkg.billingStatus})`)
    }
    if (pkg.totalDueJmd == null) throw new Error(`${pkg.trackingNumber} has no bill amount`)
    if (await getPackageCheckoutItem(pkg)) {
      throw new Error(`${pkg.trackingNumber} is already paid`)
    }
    pkgs.push(pkg)
  }

  if (!pkgs.length) throw new Error('Select at least one package to checkout')
  return pkgs
}

export async function recordPaymentCheckout(
  customer: UserRow,
  packageIds: unknown[],
  opts: {
    method: string
    recordedBy: UserRow
    reference?: string | null
    notes?: string | null
    processingFeeJmd?: number | null
  },
): Promise<CheckoutRow> {
  if (!(PAYMENT_METHODS as readonly string[]).includes(opts.method)) {
    throw new Error('Invalid payment method')
  }

  const pkgs = await validateCheckoutPackages(customer, packageIds)
  const [deliveryRequest, deliveryFee] = await resolveDeliveryRequestForPayment(customer, pkgs)

  let total = 0
  const lineAmounts: Array<[PackageRow, number]> = []

  for (const pkg of pkgs) {
    const amount = decimal(pkg.totalDueJmd)
    if (amount <= 0) throw new Error(`${pkg.trackingNumber} has an invalid bill amount`)
    total += amount
    lineAmounts.push([pkg, amount])
  }

  if (deliveryFee > 0) total += deliveryFee

  let processingFee = 0
  if (opts.processingFeeJmd != null) {
    processingFee = decimal(opts.processingFeeJmd)
    if (processingFee < 0) throw new Error('Processing fee cannot be negative')
    if (processingFee > 0) total += processingFee
  }

  const invoiceNumber = await generateInvoiceNumber()
  const methodLabel = opts.method.replace(/_/g, ' ')

  const [checkout] = await db
    .insert(paymentCheckouts)
    .values({
      customerId: customer.id,
      invoiceNumber,
      totalJmd: total.toFixed(2),
      method: opts.method,
      reference: (opts.reference ?? '').trim() || null,
      notes: (opts.notes ?? '').trim() || null,
      recordedById: opts.recordedBy.id,
      recordedAt: new Date(),
      deliveryRequestId: deliveryRequest?.id ?? null,
      deliveryFeeJmd: deliveryFee > 0 ? deliveryFee.toFixed(2) : null,
      processingFeeJmd: processingFee > 0 ? processingFee.toFixed(2) : null,
    })
    .returning()

  for (const [pkg, amount] of lineAmounts) {
    await db.insert(paymentCheckoutItems).values({
      checkoutId: checkout!.id,
      packageId: pkg.id,
      amountJmd: amount.toFixed(2),
    })
    await db
      .update(packages)
      .set({ billingStatus: 'paid', updatedAt: sql`now()` })
      .where(eq(packages.id, pkg.id))
    await addPackageEvent(
      pkg,
      pkg.status,
      `Payment recorded (${methodLabel}) — invoice ${checkout!.invoiceNumber}`,
    )
  }

  return checkout!
}

export async function recordPackagePayment(
  pkg: PackageRow,
  opts: {
    method: string
    recordedBy: UserRow
    reference?: string | null
    notes?: string | null
  },
): Promise<CheckoutRow> {
  const [customer] = await db.select().from(users).where(eq(users.id, pkg.customerId)).limit(1)
  if (!customer) throw new Error('Package customer not found')
  return recordPaymentCheckout(customer, [pkg.id], opts)
}

export async function getCheckoutItems(checkoutId: string): Promise<
  Array<{
    item: CheckoutItemRow
    pkg: PackageRow | null
  }>
> {
  const items = await db
    .select()
    .from(paymentCheckoutItems)
    .where(eq(paymentCheckoutItems.checkoutId, checkoutId))

  const result = []
  for (const item of items) {
    const [pkg] = await db.select().from(packages).where(eq(packages.id, item.packageId)).limit(1)
    result.push({ item, pkg: pkg ?? null })
  }
  return result
}

export async function getCheckoutCustomer(customerId: string): Promise<UserRow | null> {
  const [customer] = await db.select().from(users).where(eq(users.id, customerId)).limit(1)
  return customer ?? null
}
